casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -86,13 +86,47 @@ object EstimatorHelper {
       ancestors = maybeAncestors.flatten
       ancestorEvents = ancestors.flatMap(_.getBody.deploys.flatMap(_.deployLog)) ++
         ancestors.flatMap(_.getBody.deploys.flatMap(_.paymentLog))
+      commEvents = ancestorEvents.filter {
+        case Event(Comm(CommEvent(Some(_: ConsumeEvent), _))) =>
+          true
+        case _ => false
+      }
+      producesInCommEvents = commEvents.flatMap {
+        case Event(Comm(CommEvent(Some(_: ConsumeEvent), produces))) =>
```

Comment:
this repeating pattern is not nice. At line 89 you can extract the commEvent from the Event(Comm(CommEvent(_))) pattern and then partition them in produces and consumes. Then you only work with the CommEvent objects, which would be simpler

Reply:
PTAL

Reply:
Still I'm not happy with the output. I tried some refactoring on my machine and I think it looks readable and it seems to compile but I haven't run the tests. Do you think this works?

Also we could move the extractors to some common place where they could be reused.

```scala

  val extractProduce : Event => Option[ProduceEvent] = {
    case Event(Produce(produce: ProduceEvent)) => Some(produce)
    case _ => None
  }


  val extractConsume : Event => Option[ConsumeEvent] = {
    case Event(Consume(consume: ConsumeEvent)) => Some(consume)
    case _ => None
  }

  val extractComm : Event => Option[CommEvent] = {
    case Event(Comm(commEvent: CommEvent)) => Some(commEvent)
    case _ => None
  }

  private def buildBlockAncestorChannels[F[_]: Monad: BlockStore](
      blockAncestorsMeta: List[BlockMetadata]
  ): F[Set[ByteString]] =
    for {
      maybeAncestors <- blockAncestorsMeta.traverse(
                         blockAncestorMeta => BlockStore[F].get(blockAncestorMeta.blockHash)
                       )
      ancestors = maybeAncestors.flatten
      ancestorEvents = 
        ancestors.flatMap(_.getBody.deploys.flatMap(_.deployLog)) ++
        ancestors.flatMap(_.getBody.deploys.flatMap(_.paymentLog))

      produceEvents = ancestorEvents.map(extractProduce).flatten
      consumeEvents = ancestorEvents.map(extractConsume).flatten
      commEvents = ancestorEvents.map(extractComm).flatten

      producesInCommEvents = commEvents.flatMap(_.produces)
      
      consumeInCommEvents = commEvents.flatMap(_.consume)

      ancestorProduceChannels =
        ancestorEvents
          .map(extractProduce).flatten
          .filter(producesInCommEvents.contains(_))
          .map(p => p.channelsHash)
          .toSet

      ancestorConsumeChannels =
        ancestorEvents
          .map(extractConsume).flatten
          .filter(consumeInCommEvents.contains(_))
          .flatMap(c => c.channelsHashes)
          .toSet

      ancestorCommChannels =
        ancestorEvents
          .map(extractComm).flatten
          .filter(ce => consumeEvents.contains(ce.consume) && ce.produces.forall(produce => produceEvents.contains(produce)))
          .flatMap(c => c.consume.toSeq.map(_.channelsHashes) ++ c.produces.map(_.channelsHash))
          .toSet

      ancestorChannels = ancestorProduceChannels ++ ancestorConsumeChannels ++ ancestorCommChannels
    } yield ancestorChannels
```